src/txrequest.h

```diff
@@ -0,0 +1,193 @@
+ *   - If for a given txhash only already-requested announcements remain, they are all forgotten.
+ *
+ *   Rationale: giving a peer multiple chances to announce a transaction would allow them to bias requests in their
+ *              favor, worsening transaction censoring attacks. The flip side is that as long as an attacker manages
+ *              to prevent us from receiving a transaction, failed announcements (including those from honest peers)
+ *              will linger longer, increasing memory usage somewhat. The impact of this is limited by imposing a
+ *              cap on the number of tracked announcements per peer. As failed requests in response to announcements
+ *              from honest peers should be rare, this almost solely hinders attackers.
+ *              Transaction censoring attacks can be done by announcing transactions quickly while not answering
+ *              requests for them. See https://allquantor.at/blockchainbib/pdf/miller2015topology.pdf for more
+ *              information.
+ *
+ * - Transactions are not requested from a peer until its reqtime has passed.
+ *
+ *   Rationale: enable the calling code to define a delay for less-than-ideal peers, so that (presumed) better
+ *              peers have a chance to give their announcement first.
+ *
+ * - If multiple viable candidate peers exist according to the above rules, pick a peer as follows:
+ *
+ *   - If any preferred peers are available, non-preferred peers are not considered for what follows.
+ *
+ *     Rationale: preferred peers are more trusted by us, so are less likely to be under attacker control.
+ *
+ *   - Pick a uniformly random peer among the candidates.
+ *
+ *     Rationale: random assignments are hard to influence for attackers.
+ *
+ * Together these rules strike a balance between being fast in non-adverserial conditions and minimizing
+ * susceptibility to censorship attacks. An attacker that races the network:
+ * - Will be unsuccessful if all preferred connections are honest (and there is at least one preferred connection).
+ * - If there are P preferred connections of which Ph>=1 are honest, the attacker can delay us from learning
+ *   about a transaction by k expiration periods, where k ~ 1 + NHG(N=P-1,K=P-Ph-1,r=1), which has mean
+ *   P/(Ph+1) (where NHG stands for Negative Hypergeometric distribution). The "1 +" is due to the fact that the
+ *   attacker can be the first to announce through a preferred connection in this scenario, which very likely means
+ *   they get the first request.
+ * - If all P preferred connections are to the attacker, and there are NP non-preferred connections of which NPh>=1
+ *   are honest, where we assume that the attacker can disconnect and reconnect those connections, the distribution
+ *   becomes k ~ P + NB(p=1-NP/NPh,r=1) (where NB stands for Negative Binomial distribution), which has mean
+ *   P-1+NP/NPh.
```

Comment:
Random idea that I don't think we should do right now: I think we could limit this attack further if we added a `bool m_horribly_delayed : 1;` bit to each announcement. Change the ordering for choosing BEST to be `(preferred, horribly_delayed, priority)` and flip the horribly_delayed bit as part of `SetTimePoint()` when `req_time + txrequest.m_horrible_delay_time < now`, where `m_horrible_delay_time` is on the order of 10 to 60 minutes. That might also be roughly reasonable logic to do some debug logging so it's possible to catch weird behaviour in the wild.

Reply:
Nice idea. I was thinking in a similar direction, but couldn't figure out how to make it efficient. Updating the flag after the fact is a neat solution.